ARROW-9947: [Python] High-level Python API for Parquet encryption of files. (#10450)
apache/arrow: docs/source/python/parquet.rst

    abfs = AzureBlobFileSystem(account_name="XXXX", account_key="XXXX", container_name="XXXX")
    table = pq.read_table("file.parquet", filesystem=abfs)

> Parquet Modular Encryption (Columnar Encryption)

Reviewer: Thanks a lot for the detailed doc here!

------------------------------------------------

Columnar encryption is supported for Parquet files in C++ starting from
Apache Arrow 4.0.0 and in PyArrow starting from Apache Arrow 6.0.0.

Parquet uses the envelope encryption practice, where file parts are encrypted
with "data encryption keys" (DEKs), and the DEKs are encrypted with "master
encryption keys" (MEKs). The DEKs are randomly generated by Parquet for each
encrypted file/column. The MEKs are generated, stored and managed in a Key
Management Service (KMS) of user’s choice.

> Reading and writing encrypted parquet files involves passing file encryption

Reviewer: ```suggestion
Reading and writing encrypted Parquet files involves passing file encryption
```

and decryption properties to :class:`~pyarrow.parquet.ParquetWriter` and to
:class:`~.ParquetFile`, respectively.

> Writing an encrypted parquet:

Reviewer: ```suggestion
Writing an encrypted Parquet file:
```

.. code-block:: python

   encryption_properties = crypto_factory.file_encryption_properties(
                                    kms_connection_config, encryption_config)
   with pq.ParquetWriter(filename, schema,
                        encryption_properties=encryption_properties) as writer:
      writer.write_table(table)

> Reading an encrypted parquet:

Reviewer: ```suggestion
Reading an encrypted Parquet file:
```

.. code-block:: python

   decryption_properties = crypto_factory.file_decryption_properties(
                                                    kms_connection_config)
   parquet_file = pq.ParquetFile(filename,
                                 decryption_properties=decryption_properties)

> In order to create the encryption and decryption properties, a ``CryptoFactory``

Reviewer: Can you add the additional classes to `docs/source/python/api/formats.rst` (in the Parquet section), then use the proper reference markup here (probably ```:class:`~pyarrow.parquet.CryptoFactory` ```) ?

should be created and initialized with KMS Client details, as described below.

KMS Client
~~~~~~~~~~

> The master encryption keys must be kept and managed in a production-grade KMS

Reviewer: ```suggestion
The master encryption keys should be kept and managed in a production-grade KMS
```

> system, deployed in user's organization. Using Parquet encryption requires

Reviewer: Can you spell out "KMS" in full at least once? For example "Key Management System (KMS)".

provided by the user. This factory method will be used to initialize the
``CryptoFactory`` for creating file encryption and decryption properties.
For example, in order to use the ``MyKmsClient`` defined above:

.. code-block:: python

   def kms_client_factory(kms_connection_configuration):
      return MyKmsClient(kms_connection_configuration)

   crypto_factory = CryptoFactory(kms_client_factory)

An :download:`example <../../../python/examples/parquet_encryption/sample_vault_kms_client.py>`
of such a class for an open source
`KMS <https://www.vaultproject.io/api/secret/transit>`_ can be found in the Apache
Arrow GitHub repository. The production KMS client should be designed in
cooperation with an organization's security administrators, and built by
developers with experience in access control management. Once such a class is
created, it can be passed to applications via a factory method and leveraged
by general PyArrow users as shown in the encrypted parquet write/read sample
above.

KMS connection configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Configuration of connection to KMS (``kms_connection_config`` used when

> creating file encryption and decryption properties) includes the following

Reviewer: This is not very clear, what does "include" mean? Is ``kms_connection_config`` a dict? An instance of a particular type?

options:

* ``kms_instance_url``, URL of the KMS instance.
* ``kms_instance_id``, ID of the KMS instance that will be used for encryption
  (if multiple KMS instances are available).
* ``key_access_token``, authorization token that will be passed to KMS.
* ``custom_kms_conf``, a string dictionary with KMS-type-specific configuration.

Encryption configuration
~~~~~~~~~~~~~~~~~~~~~~~~

Encryption configuration (``encryption_config`` used when creating file
encryption properties) includes the following options:

* ``footer_key``, ID of the master key for footer encryption/signing.

> * ``column_keys``, list of columns to encrypt, with master key IDs.

Reviewer: What is this, a list of tuples?

> * ``uniform_encryption``, encrypt footer and all columns with the same

Reviewer: ```suggestion
* ``uniform_encryption``, whether to encrypt the footer and all columns with the same
```

  encryption key.

> * ``encryption_algorithm``, parquet encryption algorithm.

Reviewer: ```suggestion
* ``encryption_algorithm``, the Parquet encryption algorithm.
```

  Can be ``AES_GCM_V1`` (default), or ``AES_GCM_CTR_V1``.

> * ``plaintext_footer``, write files with plaintext footer.

Reviewer: ```suggestion
* ``plaintext_footer``, whether to write the file footer in plain text (otherwise it is encrypted).
```

> * ``double_wrapping``, use double wrapping - where data encryption keys (DEKs)

Reviewer: ```suggestion
* ``double_wrapping``, whether to use double wrapping - where data encryption keys (DEKs)
```

  are encrypted with key encryption keys (KEKs), which in turn are encrypted

>   with master keys. If set to false, use single wrapping - where DEKs are

Reviewer: ```suggestion
  with master keys. If set to ``false``, single wrapping is used - where DEKs are
```

  encrypted directly with master keys.
* ``cache_lifetime``, lifetime of cached entities (key encryption keys,

>   local wrapping keys, KMS client objects)

Reviewer: In which unit is this expressed?

> * ``internal_key_material``, store key material inside Parquet file footers;

Reviewer: ```suggestion
* ``internal_key_material``, whether to store key material inside Parquet file footers;
```

>   this mode doesn’t produce additional files. If set to false, key material is

Reviewer: ```suggestion
  this mode doesn’t produce additional files. If set to ``false``, key material is
```

  stored in separate files in the same folder, which enables key rotation for

>   immutable Parquet files.

Reviewer: For the record, by "key material", we mean the MEK-encrypted DEKs and KEKs? The DEKs don't change, but their encrypted storage might due to rotation of KEKs and/or MEKs?

Author: You're right. DEKs don't change, because this would mean encrypting the Parquet files entirely again. The new MEKs (and transparently re-generated KEKs) are used to re-encrypt the DEKs.

* ``data_key_length_bits``, length of data encryption keys (DEKs), randomly

>   generated by parquet key management tools. Can be 128, 192 or 256 bits.

Reviewer: ```suggestion
  generated by Parquet key management tools. Can be 128, 192 or 256 bits.
```

.. note::
   By default, Parquet implements a "double envelope encryption" mode, that
   minimizes the interaction of the program with a KMS server. In this mode,
   the DEKs are encrypted with "key encryption keys" (KEKs, randomly generated
   by Parquet). The KEKs are encrypted with MEKs in KMS; the result and the
   KEK itself are cached in the process memory. Users interested in regular
   envelope encryption, can switch to it by setting the double_wrapping

>    parameter of EncryptionConfiguration to false.

Reviewer: Can you use the proper Sphinx markup to link to the EncryptionConfiguration API here?

An example encryption configuration:

.. code-block:: python

   encryption_config = pq.EncryptionConfiguration(
      footer_key="footer_key_name",
      column_keys={
         "column_key_name": ["Column1", "Column2"],
      },
   )

Decryption configuration
~~~~~~~~~~~~~~~~~~~~~~~~
   
Decryption configuration (``decryption_config`` used when creating file

> decryption properties) is optional and it includes the following options:

Reviewer: Similar question as for encryption configuration.

* ``cache_lifetime``, lifetime of cached entities (key encryption keys, local

>   wrapping keys, KMS client objects).

Reviewer: In which unit is the lifetime expressed?

Author: It's a datetime.timedelta. Adding it to the description.

                                                    kms_connection_config)
   parquet_file = pq.ParquetFile(filename,
                              decryption_properties=decryption_properties)

In order to create the encryption and decryption properties, a ``CryptoFactory``
should be created and initialized with KMS Client details, as described below.

KMS Client
~~~~~~~~~~

The master encryption keys must be kept and managed in a production-grade KMS
system, deployed in user's organization. Using Parquet encryption requires
implementation of a client class for the KMS server.
Any KmsClient implementation should implement the following informal interface:

.. code-block:: python

   class KmsClient:
   def wrap_key(self, key_bytes, master_key_identifier):
      """Wrap a key - encrypt it with the master key."""
       raise NotImplementedError()

   def unwrap_key(self, wrapped_key, master_key_identifier):
      """Unwrap a key - decrypt it with the master key."""
      raise NotImplementedError()

An example KmsClient impelementation might look like the following:

.. code-block:: python
   class MyKmsClient(pq.KmsClient):
      def __init__(self, kms_connection_configuration):

>          pq.KmsClient.__init__(self)

Reviewer 2: The subclassing from pq.KmsClient is required, right? In that case, I would also add it to the minimal code snippet laying out the required methods above (or maybe combine the two code blocks into one)
